### PR TITLE
layout: Add support for table rows, columns, rowgroups and colgroups

### PR DESCRIPTION
**Pull request by mrobinson**
This adds support for table rows, columns, rowgroups and colgroups.
There are few additions here:

1. The creation of fragments, which allows script queries and hit
   testing to work properly. These fragments are empty as all cells are
   still direct descendants of the table fragment.
2. Properly handling size information from tracks and track groups as
   well as frustrating rules about reordering rowgroups.
3. Painting a background seemlessly across track groups and groups. This
   is a thing that isn't done in legacy layout (nor WebKit)!

This causes some tests to start failing, because it exposes our lack of
support for collapse in tables.

Co-authored-by: Oriol Brufau <obrufau@igalia.com>

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

### CODE REVIEW

File: components/layout_2020/display_list/mod.rs [hunk 1]
```diff
@@ -205,7 +206,21 @@ impl Fragment {
...
-            Fragment::Anonymous(_) => {},
+            Fragment::Positioning(positioning_fragment) => {
+                if let Some(style) = positioning_fragment.style.as_ref() {
+                    let rect = positioning_fragment
+                        .rect
+                        .to_physical(style.writing_mode, containing_block)
+                        .translate(containing_block.origin.to_vector());
+                    self.maybe_push_hit_test_for_style_and_tag(
+                        builder,
+                        &style,
```
Comment (Loirooriol) [inline components/layout_2020/display_list/mod.rs:217]:
```suggestion
                        style,
```

File: components/layout_2020/display_list/mod.rs [hunk 2]
```diff
@@ -550,34 +605,36 @@ impl<'a> BuilderForBoxFragment<'a> {
...
+        };
+        self.build_background_for_painter(painter, builder);
     }
 
-    fn build_background_image(
+    fn build_background_image<'b>(
         &mut self,
         builder: &mut DisplayListBuilder,
-        source: background::Source<'a>,
+        painter: BackgroundPainter<'b>,
```
Comment (Loirooriol) [inline components/layout_2020/display_list/mod.rs:634]:
```diff
-    fn build_background_image<'b>(
+    fn build_background_image(
         &mut self,
         builder: &mut DisplayListBuilder,
-        painter: BackgroundPainter<'b>,
+        painter: BackgroundPainter<'_>,
```

File: components/layout_2020/display_list/mod.rs [hunk 3]
```diff
@@ -540,6 +574,27 @@ impl<'a> BuilderForBoxFragment<'a> {
         builder.wr().push_hit_test(&common, hit_info);
     }
 
+    fn build_background_for_painter<'b>(
+        &mut self,
+        painter: BackgroundPainter<'b>,
+        builder: &mut DisplayListBuilder,
```
Comment (Loirooriol) [inline components/layout_2020/display_list/mod.rs:580]:
```suggestion
    fn build_background_for_painter(
        &mut self,
        painter: BackgroundPainter<_>,
        builder: &mut DisplayListBuilder,
```

File: components/layout_2020/fragment_tree/positioning_fragment.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    ) -> Self {
+        Self::new_with_base_fragment(BaseFragment::anonymous(), None, rect, children, mode)
+    }
+
+    pub fn new_empty(
+        base_fragment_info: BaseFragmentInfo,
+        rect: LogicalRect<Length>,
+        style: ServoArc<ComputedValues>,
+    ) -> Self {
+        let writing_mode = style.writing_mode.clone();
```
Comment (Loirooriol) [inline components/layout_2020/fragment_tree/positioning_fragment.rs:48]:
```suggestion
        let writing_mode = style.writing_mode;
```

File: components/layout_2020/table/layout.rs [hunk 5]
```diff
@@ -132,84 +136,47 @@ impl<'a> TableLayout<'a> {
...
-                    .style
-                    .max_box_size(writing_mode)
-                    .inline
-                    .and_then(|length_percentage| length_percentage.to_percentage())
-                    .unwrap_or(Percentage(f32::INFINITY));
-                let percentage_contribution =
-                    Percentage(inline_size_percent.0.min(max_inline_size_percent.0));
+                let pbm = cell.style.padding_border_margin(containing_block);
+                outer_min_content_width += pbm.padding_border_sums.inline.into();
+                outer_max_content_width += pbm.padding_border_sums.inline.into();
```
Comment (Loirooriol) [inline components/layout_2020/table/layout.rs:172]:
```suggestion
                outer_min_content_width += pbm.padding_border_sums.inline;
                outer_max_content_width += pbm.padding_border_sums.inline;
```

File: components/layout_2020/table/layout.rs [hunk 6]
```diff
@@ -1078,46 +1052,65 @@ impl<'a> TableLayout<'a> {
...
+                let cell_rect = dimensions.get_cell_rect(
+                    TableSlotCoordinates::new(column_index, row_index),
+                    cell.rowspan,
+                    cell.colspan,
+                );
+
                 // If this cell has baseline alignment, it can adjust the table's overall baseline.
+                let row_baseline = self.row_baselines[row_index];
                 if cell.effective_vertical_align() == VerticalAlignKeyword::Baseline {
+                    let baseline = Au::from(cell_rect.start_corner.block) + row_baseline;
```
Comment (Loirooriol) [inline components/layout_2020/table/layout.rs:1064]:
```suggestion
                    let baseline = cell_rect.start_corner.block + row_baseline;
```

**Review by Loirooriol — COMMENTED** [review]
Some complaints from clippy

**Comment by mrobinson** [comment]
Thanks @Loirooriol, I think I've fixed those clippy warnings, but there are so many in Layout 2020 it's a bit hard to tell. It would be nice to clean them all up.

**Review by Loirooriol — COMMENTED** [review]
I think I didn't have time to review all the files, I will go through it again, but for now I guess I can submit the comments that I had.

**Review by Loirooriol — APPROVED** [review]
The legacy expectations for the html/dom tests need to be updated.

**Pull request merged**